hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisUnderReplicationHandler.java

```diff
@@ -0,0 +1,247 @@
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Returns a list of datanodes that can be used as sources for replication
+   * for the container specified in replicaCount.
+   *
+   * @param replicaCount RatisContainerReplicaCount object for this container
+   * @param pendingOps List of pending ContainerReplicaOp
+   * @return List of healthy datanodes that have closed/quasi-closed replicas
+   * and are not pending replica deletion. Sorted in descending order of
+   * sequence id.
+   */
+  private List<DatanodeDetails> getSources(
+      RatisContainerReplicaCount replicaCount,
+      List<ContainerReplicaOp> pendingOps) {
+    Set<DatanodeDetails> pendingDeletion = new HashSet<>();
+    // collect the DNs that are going to have their container replica deleted
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        pendingDeletion.add(op.getTarget());
+      }
+    }
+
+    /*
+     * Return healthy datanodes that have closed/quasi-closed replicas and
+     * are not pending replica deletion. Sorted in descending order of
+     * sequence id.
+     */
+    return replicaCount.getReplicas().stream()
+        .filter(r -> r.getState() == State.QUASI_CLOSED ||
+            r.getState() == State.CLOSED)
+        .filter(r -> ReplicationManager.getNodeStatus(r.getDatanodeDetails(),
+            nodeManager).isHealthy())
+        .filter(r -> !pendingDeletion.contains(r.getDatanodeDetails()))
+        .sorted((r1, r2) -> r2.getSequenceId().compareTo(r1.getSequenceId()))
+        .map(ContainerReplica::getDatanodeDetails)
+        .collect(Collectors.toList());
+  }
+
+  private List<DatanodeDetails> getTargets(
+      RatisContainerReplicaCount replicaCount,
+      List<ContainerReplicaOp> pendingOps) throws IOException {
+    // DNs that already have replicas cannot be targets and should be excluded
+    final List<DatanodeDetails> excludeList =
+        replicaCount.getReplicas().stream()
+            .map(ContainerReplica::getDatanodeDetails)
+            .collect(Collectors.toList());
+
+    // DNs that are already waiting to receive replicas cannot be targets
+    final List<DatanodeDetails> pendingReplication =
+        pendingOps.stream()
+            .filter(containerReplicaOp -> containerReplicaOp.getOpType() ==
+                ContainerReplicaOp.PendingOpType.ADD)
+            .map(ContainerReplicaOp::getTarget)
+            .collect(Collectors.toList());
+    excludeList.addAll(pendingReplication);
+
+    /*
+    Ensure that target datanodes have enough space to hold a complete
+    container.
+    */
+    final long dataSizeRequired =
+        Math.max(replicaCount.getContainer().getUsedBytes(),
+            currentContainerSize);
+    return placementPolicy.chooseDatanodes(excludeList, null,
+        replicaCount.additionalReplicaNeeded(), 0, dataSizeRequired);
+
+    // TODO Now that under replication handling is separate from mis
```

Comment:
We probably need to review how the placement policy works, but I think what is *should* do is:

1. If there are multiple racks online, it must either give a node which meets the policy
2. Or, throw SCMException with insufficient nodes, which means we cannot recover now.
3. If there are not enough racks online, eg a 2 rack cluster and one rack is offline, then it should give a node which violates the policy. This is because fixing under replication is more important than causing mis-replication.

Its not a great design if we have to check the node we are given actually meets the policy here again.

An alternative to the above, is to have an "allowed to voilate policy" flag we can pass into the policy. That way, when we are choosing nodes it will always throw if it cannot find a node that meets the policy, and we can catch that and try again with the flag set to true, but that may be overkill.

Reply:
> If there are not enough racks online, eg a 2 rack cluster and one rack is offline, then it should give a node which violates the policy. This is because fixing under replication is more important than causing mis-replication.

Checked that this behaviour is allowed if `fallback` is enabled in the placement policy. It's true for the ratis placement policy that's constructed in SCM, so we should be good.
```
    containerPlacementPolicy =
        ContainerPlacementPolicyFactory.getPolicy(conf, scmNodeManager,
            clusterMap, true, placementMetrics);

```